### DISCUSSION TITLE
Using Terragrunt for large scale disaster recovery

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#110] [Help] [answered]
Do you have any examples or case studies on using Terragrunt to do a large scale disaster recovery operation?

For example: Deploying 100+ resources to a new cloud region in the event of a disaster in another region

What is your recommendation for how to use Terragrunt for such an event?

**Comment by gruntwork-discussions** [comment]
You must label this issue with at least one of the following labels before it can be answered:

* [r:terraform-aws-service-catalog](https://github.com/gruntwork-io/terraform-aws-service-catalog)
* [r:terraform-aws-vpc](https://github.com/gruntwork-io/terraform-aws-vpc)
* [r:terraform-aws-architecture-catalog](https://github.com/gruntwork-io/terraform-aws-architecture-catalog)
* [r:terraform-aws-asg](https://github.com/gruntwork-io/terraform-aws-asg)
* [r:terraform-aws-beanstalk](https://github.com/gruntwork-io/terraform-aws-beanstalk)
* [r:terraform-aws-cache](https://github.com/gruntwork-io/terraform-aws-cache)
* [r:terraform-aws-ci](https://github.com/gruntwork-io/terraform-aws-ci)
* [r:terraform-aws-ci-pipeline-example](https://github.com/gruntwork-io/terraform-aws-ci-pipeline-example)
* [r:terraform-aws-cis-service-catalog](https://github.com/gruntwork-io/terraform-aws-cis-service-catalog)
* [r:terraform-aws-data-storage](https://github.com/gruntwork-io/terraform-aws-data-storage)
* [r:terraform-aws-ecs](https://github.com/gruntwork-io/terraform-aws-ecs)
* [r:terraform-aws-eks](https://github.com/gruntwork-io/terraform-aws-eks)
* [r:terraform-aws-elk](https://github.com/gruntwork-io/terraform-aws-elk)
* [r:terraform-aws-kafka](https://github.com/gruntwork-io/terraform-aws-kafka)
* [r:terraform-aws-lambda](https://github.com/gruntwork-io/terraform-aws-lambda)
* [r:terraform-aws-load-balancer](https://github.com/gruntwork-io/terraform-aws-load-balancer)
* [r:terraform-aws-messaging](https://github.com/gruntwork-io/terraform-aws-messaging)
* [r:terraform-aws-mongodb](https://github.com/gruntwork-io/terraform-aws-mongodb)
* [r:terraform-aws-monitoring](https://github.com/gruntwork-io/terraform-aws-monitoring)
* [r:terraform-aws-openvpn](https://github.com/gruntwork-io/terraform-aws-openvpn)
* [r:terraform-aws-sam](https://github.com/gruntwork-io/terraform-aws-sam)
* [r:terraform-aws-security](https://github.com/gruntwork-io/terraform-aws-security)
* [r:terraform-aws-server](https://github.com/gruntwork-io/terraform-aws-server)
* [r:terraform-aws-static-assets](https://github.com/gruntwork-io/terraform-aws-static-assets)
* [r:terraform-aws-zookeeper](https://github.com/gruntwork-io/terraform-aws-zookeeper)
* [r:terragrunt](https://github.com/gruntwork-io/terragrunt)
* [r:terratest](https://github.com/gruntwork-io/terratest)
* [r:repo-copier](https://github.com/gruntwork-io/repo-copier)


[Community Guidelines](https://github.com/gruntwork-io/knowledge-base/discussions/1)
[Getting help](https://gruntwork.io/support/)
[Gruntwork documentation](https://www.gruntwork.io)

**Comment by brikis98** [comment, marked as answer]
We can't provide advice on disaster recovery in general, as it will be VERY use case specific. E.g., What data do you have? How are you replicating it? What sort of consistency requirements do you have? How do you handle DNS fail over? And so on.

All we can comment on here is, in relation to Terragrunt, if you have an environment `env1` deployed, how can you structure your code so it's easy to quickly spin up a duplicate environment `env2` (e.g., in case of disaster recovery)? Here's the rough idea:

1. Create `env1` using Terragrunt (i.e., folders + `terragrunt.hcl` files), with each `source` URL pointing to a specific version of your modules.
2. Ensure that any parameters that must be set differently in each environment (e.g., domain names, CIDR blocks) are extracted into top-level vars, such as e.g., an `env.hcl` file at the root of the environment. 
3. When you need to spin up an `env2`, create a copy of the `env1` folder named `env2`.
4. Go into the `env.hcl` file and update it accordingly.
5. Take care of any manual steps that are necessary before spinning up the new environment. These depend on your use case, but it's common to have certain things managed outside of your infrastructure code, by design: e.g., writing secrets to your secrets store (e.g., AWS Secrets Manager or Vault); buying a new domain name; etc.
6. Run `terragrunt run-all apply`.
